README.md
Coins
=====

This repository contains utilities to read HM Treasury COINS (Combined Online Information System) files, and to convert them to more usable formats.


Installation
------------

Install the [git](http://git-scm.com/download) version control system

Download this repository:

    git clone git@github.com:martinbudden/coins.git

Install python, if you do not already have it.


Usage
-----

Before running any of these utilities the HM Treasury COINS files should be downloaded, unzipped and placed in the "data" directory. After downloading and unzipping the COINS data released on 4th June 2010 the "data" directory should include the four files:

* fact_table_extract_2009_10.txt
* fact_table_extract_2008_09.txt
* adjustment_table_extract_2009_10.txt
* adjustment_table_extract_2008_09.txt

These files are in unicode format and need to be converted to UTF-8 format for easier usage. Run coins2utf8.py to do this:

    python coins2utf8.py

This currently only operates on the fact_table_extract_*.txt files. It currently does three operations:

* converts the file to UTF-8 csv format (field delimiter is still '@')
* extracts only those records where the 'Value' field is non-zero

The output of coins2utf8.py is the files facts_2008_09_nz.csv and facts_2009_10_nz.csv. Note that 'nz' stands for 'non-zero'. The file facts_2009_10_s1000.csv is also produced - this is a small file of 1000 records which can be loaded into a spreadsheet to get a feel for the data. (Note that the field delimiter is '@' and this needs to be specified when loading the file, not also that one of the departmental_codes is 'DEC066', which may be interpreted as a date - be sure to set the 'Column type' for this column to be 'TEXT' when loading the file into a spreadsheet.)

Next run coinsdescriptions.py. This extracts varies COINS codes and descriptions (eg department_code and department_description) and places them in .csv files the data/desc directory. These files are standard .csv files and can be loaded into (say) Excel.

    python coinsdescriptions.py

This completes the preliminary preprocessing of the coins files. You may now run coins2csv.py to produce .csv files that can be directly loaded into Excel, or coins2sqlite.py which produces an Sqlite database of the coins data.


###Producing COINS files that can be loaded into Excel

Run coins2csv.py. This uses the the facts_*_nz.csv files output from coins2utf8.py to produce a .csv file for the data_types 'outturn', 'forecasts', 'plans' and 'snapshots'.

    python coins2csv.py

Most of these .csv files are small enough (less than 65,535 rows) to be loaded into Excel. These files are in the data/csv directory, and are named: 

    coin_<date_period>_<data_type>_nz.csv, eg coins_2009_10_outturn_nz.csv.


###Producing a COINS SQLite database

Run coins2sqlite.py. This uses the the facts_*_.csv files output from coins2utf8.py and the description .csv files produced by coinsdescription.py, to produce an SQLite database that can be queried.

    python coins2sqlite.py

This produces the files coins_2008_09_sqlite.db and coins_2009_10_sqlite.db in the data/sqlite directory

###Example SQL queries

Run coinssqlexamples.py. This executes a number of example queries on the COINS SQLite datbase, coins_2009_10_sqlite.db.

    python coinssqlexamples.py

###Recreating government PESA reports

Run coinssqlpesa.py. This produces HTML tables of the 2008-09 government departmental group spending figures. The tables include two empty columns than can be manually filled with the actual PESA data, so that the figures can be compared.

    python coinssqlpesa.py

Data Source
-----------

* [HM Government Combined Online Information System](http://data.gov.uk/dataset/coins)
* [HM Treasury - Understanding the COINS data](http://www.hm-treasury.gov.uk/d/coins_guidance_040610.pdf)


Treasury Information
--------------------

HM Treasury publish documents that may help to understand COINS and, and how that data may be used and aggregated. These include:

* [Public spending planning and control - a brief introduction](http://www.hm-treasury.gov.uk/psr_spend_plancontrol.htm)
* [Consolidated Budgeting Guidance](http://www.hm-treasury.gov.uk/psr_bc_consolidated_budgeting.htm)
* [Classification papers](http://www.hm-treasury.gov.uk/psr_bac_classification_papers.htm)
* Information on the Supply Estimates process at: [Financial reporting - Parliamentary Supply Estimates](http://www.hm-treasury.gov.uk/psr_estimates_index.htm)

Slightly differing approaches are taken to some data recording and outputs for the Devolved Administrations. Some information on that, by means of concordats, is available here: [Devolved Assemblies](http://www.hm-treasury.gov.uk/psr_devolved_assemblies.htm)

HM Treasury publishes a number of documents based COINS data. Public Spending Statistics are available here: [Statistics on Public Finance and Spending](http://www.hm-treasury.gov.uk/finexp_index.htm).


Checking Data
-------------

[PESA 2010 section 1 - Budgets](http://www.hm-treasury.gov.uk/pesa2010_section1.htm) gives the outrun DEL, AME, Capital and Resource outrun figures for 2004-05 to 2008-09. Outrun figures from the COINS 2008/09 Fact table can be checked against these tables.

[Public Expenditure Statistical Analyses 2009](http://www.hm-treasury.gov.uk/pespub_pesa09.htm) gives the outrun DEL, AME, Capital and Resource outrun figures for 2003-04 to 2007-08 and the estimated outrun for 2008-09 and the planned outruns for 2009-10 and 2010-2011. Forecast figures from the COINS 2009/10 Fact table can be checked against these tables.

I've briefly outlined what some of the fields in the COINS data mean in the [data/README.md](http://github.com/martinbudden/coins/blob/master/data/README.md) file.


Useful links
------------

* [Where Does My Money Go?](http://www.wheredoesmymoneygo.org/)
* Where Does My Money Go? - [A User Guide to COINS](http://www.wheredoesmymoneygo.org/data/coins/)
* Open Knowledge Foundation Blog - [The Hunt For COINS](http://blog.okfn.org/2010/02/22/the-hunt-for-coins/)
* Open Knowledge Foundation Blog - [COINS: A Users Guide](http://blog.okfn.org/2010/06/04/coins-a-users-guide/)
* What Do They Know [COINS database schema](http://www.whatdotheyknow.com/request/25039/response/67260/attach/3/100111%20COINS%20Schema%20for%20FOI%209%201049.xls)
* EtherPad - [Coins open notepad](http://pad.okfn.org/coins)
* Google spreadsheet of [COINS Schema overview:](http://spreadsheets.google.com/ccc?key=0Ah8UkI7xG7eWdHpYMnhaWmR5NVdNUG9yTkNfQVlUTWc&hl=en_GB)


Coding resources
----------------

* Open Knowledge Foundation(okfn) [coins python scripts](http://bitbucket.org/okfn/coins)


CHANGELOG
0.1 Sat Jun 05 13:00:00 BST 2010

    * Initial commit


coins/coinssqlpesa.py
#!/usr/bin/env python
#coding=utf-8
#file: coinssqlpesa.py

"""
Examples of SQL queries on the COINS database.
"""

import csv
import sqlite3
import locale
from jinja2 import Template
from collections import defaultdict
from optparse import OptionParser

# Taken from: "Public Expenditure Statistical Analysis 2009, Appendix B: Departmental Groups"
DEPARTMENTAL_GROUPS_2008_09 = [
    ('Children, Schools and Families', ('DES022', 'OIS022')),
    ('Health', ('DOH033', 'DPM085')),
    ('Transport', ('DFT004', 'ORR088')),
    ('Innovation, Universities and Skills', ('DIS063',)),
    ('CLG Communities and Local Government', ('DPM085',)),
    ('Home Office', ('HOF034',)), # missing 'Assets Recovery Agency'
    ('Justice', ('LCD047', 'PRO067', 'ELC011', 'NIC081', 'LRG045', 'SCO042', 'WOF091', 'TSC068')),
    ('Law Officers Departments', ('CPS016', 'SFO019', 'HGT089', 'RCP046')),
    ('Defence', ('MOD017',)),
    ('Foreign and Commonwealth Office', ('FCO027',)),
    ('International Development', ('DID030',)),
    ('Energy and Climate Change', ('DEC066', 'OGE020')),
    ('Business, Enterprise and Regularity Reform', ('BTI013', 'OFT074', 'PSC006', 'ECG025')), # missing 'Department for Business, Enterprise and Regulartory Reform', 'Office of Communications'
    ('Environment, Food and Rural Affairs', ('EFR003', 'OWS057')), # missing 'Forrestry Commision'
    ('Culture, Media and Sport', ('DCM048',)),
    ('Work and Pensions', ('DWP032', 'GEO064')),
    ('Scotland', ('SCT075',)), # used Scottish Government
    ('Wales', ('WAG090',)),
    ('Northern Ireland Executive', ('NIE099',)),
    ('Northern Ireland Office', ('NIO097',)),
    ('Chancellors Departments', ('HMT087', 'DNS049', 'GAD031', 'ILR041', 'CES014')), # missing 'National Investment and Loans Office', 'Royal Mint', and 'Office of Government Commerce'
    ('Cabinet Office', ('CAB010', 'COI010', 'CHC009', 'NSG062')), #missing 'Security and Intelligence Agencies'
    ('Independent Bodies', ('HOC036', 'HOL037', 'NAO050', 'ONS005', 'OPC060')) #used 'Office for National Statistics' instead of 'Statistics Board'
    ]


def table_html(data, caption, rowheaders, colheaders):
    """
    Format data into an HTML table

    Keyword arguments:
    data -- a 2D dict of the data to be formatted
    caption -- the table's caption
    rowheaders -- the table's row headings
    colheaders -- the table's column headings

    """
    template = Template('<table>\n'
        '<caption>{{caption}}</caption>\n'
        '<thead>\n'
        '<tr> '
        '<th></th> '
        '{% for colheader in colheaders %}'
            '<th>{{colheader|replace(" ", "<br />")}}</th> '
        '{% endfor %}'
        '</tr>\n'
        '</thead>\n'
        '<tbody>\n'
        '{% for rowheader in rowheaders %}'
            '<tr> '
            '<th>{{rowheader}}</th> '
            '{% for colheader in colheaders %}'
                '<td>{{data[colheader][rowheader]|escape}}</td> '
            '{% endfor %}'
            '</tr>\n'
        '{% endfor %}'
        '</tbody>\n'
        '</table>\n')
    return template.render(data=data, caption=caption, rowheaders=rowheaders, colheaders=colheaders)



def sqlite_totals(sqlite_filename):
    """
    Perform SQL queries to calucate various Resource, Capital, DEL and AME
    spending totals.

    Keyword arguments:
    sqlite_filename -- name of SQLite database file to be queried.

    """
    connection = sqlite3.connect(sqlite_filename)
    cursor = connection.cursor()

    print "CHC009 DEL outturn"
    dept = 'CHC009'
    cursor.execute('''SELECT Department_code, department.description,
        Value,
        data_subtype,
        Account_code, account.description,
        Counterparty_code,
        Time, Budget_Boundary,
        Resource_Capital,
        Dept_Group,
        CGA_Body_Type,
        COFOG,
        PESA,
        Territory
        FROM outturn
        JOIN department
        ON department.code=outturn.Department_code
        JOIN account
        ON account.code=outturn.Account_code
        WHERE Data_type=:data and Department_code=:dept and Budget_Boundary=:budget and Resource_Capital="Capital"
        and (Data_subtype LIKE "%Approved%" or Data_subtype="Submitted (Outturn)")
        ORDER BY Department_code''',
        {'data': 'Outturn', 'dept': dept, 'budget': 'DEL'})
    print cursor.fetchone()
    for row in cursor:
        print row
    print

    print "CHC009 DEL outturn TOTAL"
    cursor.execute('''SELECT SUM(Value)
        FROM outturn
        WHERE Department_code=:dept and Budget_Boundary=:budget and Resource_Capital="Capital"
        and (Data_subtype LIKE "%Approved%" or Data_subtype="Submitted (Outturn)")''',
        {'dept': dept, 'budget': 'DEL'})
    print 'TOTAL:', cursor.fetchone()[0]
    print

    locale.setlocale(locale.LC_ALL, "")
    cursor.execute('''SELECT SUM(Value)
        FROM outturn
        WHERE Budget_Boundary="DEL" and Resource_Capital="Capital"
        and (Data_subtype LIKE "%Approved%" or Data_subtype="Submitted (Outturn)")''')
    print 'TOTAL DEL Capital: ', locale.format('%d', cursor.fetchone()[0], True)

    cursor.execute('''SELECT SUM(Value)
        FROM outturn
        WHERE Budget_Boundary="AME" and Resource_Capital="Capital"
        and (Data_subtype LIKE "%Approved%" or Data_subtype="Submitted (Outturn)")''')
    print 'TOTAL AME Capital: ', locale.format('%d', cursor.fetchone()[0], True)

    cursor.execute('''SELECT SUM(Value)
        FROM outturn
        WHERE Budget_Boundary="DEL" and Resource_Capital="Resource"
        and (Data_subtype LIKE "%Approved%" or Data_subtype="Submitted (Outturn)")''')
    print 'TOTAL DEL Resource:', locale.format('%d', cursor.fetchone()[0], True)

    cursor.execute('''SELECT SUM(Value)
        FROM outturn
        WHERE Budget_Boundary="AME" and Resource_Capital="Resource"
        and (Data_subtype LIKE "%Approved%" or Data_subtype="Submitted (Outturn)")''')
    print 'TOTAL AME Resource:', locale.format('%d', cursor.fetchone()[0], True)
    print

    cursor.close()
    connection.close()


def sqlite_department_totals(sqlite_filename, date):
    """
    Perform SQL queries to calucate Capital DEL department
    spending totals.

    Keyword arguments:
    sqlite_filename -- name of SQLite database file to be queried.
    date -- reporting period (eg 2008_09)

    """
    connection = sqlite3.connect(sqlite_filename)
    cursor = connection.cursor()

    locale.setlocale(locale.LC_ALL, "")
    budget = 'DEL'
    rc = 'Capital'
    
    depts = []
    csv_reader = csv.reader(open('../data/desc/department_%s.csv' % date, 'r'))
    # skip the column headings
    row = csv_reader.next()
    try:
        while True:
            row = csv_reader.next()
            depts.append({'code': row[0], 'desc': row[1]})
    except StopIteration:
        pass

    for i in depts:
        dept = i['code']
        desc = i['desc']
        cursor.execute('''SELECT SUM(Value)
            FROM outturn
            WHERE Budget_Boundary=:budget and Resource_Capital=:rc
            and Department_code=:dept
            and (Data_subtype LIKE "%Approved%" or Data_subtype="Submitted (Outturn)")''',
            {'dept': dept, 'budget': budget, 'rc': rc})
        total = cursor.fetchone()[0]
        if total:
            print 'Dept %(dept)s (%(desc)s) %(budget)s %(rc)s: ' % {'dept': dept, 'desc': desc, 'budget': budget, 'rc': rc},
            print locale.format('%d', total, True)

    cursor.close()
    connection.close()


def sqlite_departmental_group_totals(sqlite_filename, groups, resource_capital, budget_boundary, verbose):
    """
    Perform SQL queries to calucate department group spending totals.

    Keyword arguments:
    sqlite_filename -- name of SQLite database file to be queried.
    groups -- list of departmental groups
    resource_capital -- 'Resource' or 'Capital' for resource or capital spending
    budget_boundary -- 'DEL' or 'AME' for DEL or AME spending
    verbose -- give detailed status updates.

    """
    connection = sqlite3.connect(sqlite_filename)
    cursor = connection.cursor()

    table = defaultdict(dict)
    row_headers = []
    column_headers = []
    col = 'COINS 2008-09 outturn'
    column_headers.append(col)
    col2009 = 'PESA 2009 estimate'
    col2010 = 'PESA 2010 outturn'
    column_headers.append(col2009)
    column_headers.append(col2010)
    
    locale.setlocale(locale.LC_ALL, "")
    
    total = 0
    caption = '%s %s by departmental group (millions)' % (resource_capital, budget_boundary)
    if verbose:
        print caption
    for i in groups:
        group = i[0]
        depts = i[1]
        cursor.execute('''SELECT SUM(Value)
            FROM outturn
            WHERE Budget_Boundary=? and Resource_Capital=?
            and Department_code IN (%s)
            and (Data_subtype LIKE "%%Approved%%" or Data_subtype="Submitted (Outturn)")
            ''' % ('?,' * len(depts))[:-1],
            (budget_boundary, resource_capital) + depts)
        subtotal = cursor.fetchone()[0]
        if not subtotal:
            subtotal = 0
        table[col][group] = locale.format('%d', round(subtotal/1000), True)
        table[col2009][group] = 0
        table[col2010][group] = 0
        row_headers.append(group)
        if verbose:
            print '%(group)s: ' % {'group': group}, locale.format('%d', round(subtotal/1000), True)
        total += subtotal
    if verbose:
        print 'Total', resource_capital, budget_boundary, locale.format('%d', round(total/1000), True)
        print
    html = table_html(table, caption, row_headers, column_headers)
    print html

    cursor.close()
    connection.close()
    return table


def process_options(arglist=None):
    """
    Process options passed either via arglist or via command line args.

    """
    parser = OptionParser(arglist)
    #parser.add_option("-f", "--file", dest="filename",
    #                  help="file to be converted", metavar="FILE")
    parser.add_option("-v", action="store_true", dest="verbose", default=False,
                      help="print status messages to stdout")

    (options, args) = parser.parse_args()
    return options, args


def main():
    """
    Perform example queries on the COINS Sqlite database.

    """
    (options, args) = process_options()
    if len(args) == 0:
        sqlite_filename = '../data/sqlite/coins_2008_09_sqlite.db'
    else:
        sqlite_filename = args[0]
    sqlite_departmental_group_totals(sqlite_filename, DEPARTMENTAL_GROUPS_2008_09, 'Resource', 'DEL', options.verbose)
    sqlite_departmental_group_totals(sqlite_filename, DEPARTMENTAL_GROUPS_2008_09, 'Resource', 'AME', options.verbose)
    sqlite_departmental_group_totals(sqlite_filename, DEPARTMENTAL_GROUPS_2008_09, 'Capital', 'DEL', options.verbose)
    sqlite_departmental_group_totals(sqlite_filename, DEPARTMENTAL_GROUPS_2008_09, 'Capital', 'AME', options.verbose)


if __name__ == "__main__":
    main()
